src/ugeneui/src/main_window/CreateDesktopShortcutTask.cpp

```diff
@@ -0,0 +1,193 @@
+    runOnStartup = startUp;
+    setVerboseLogMode(true);
+    startError = false;
+}
+
+void CreateDesktopShortcutTask::run() {
+    stateInfo.setDescription(tr("Create desktop shortcut running"));
+
+    stateInfo.setDescription(QString());
+}
+
+bool CreateDesktopShortcutTask::createDesktopShortcut() {
+#if defined(Q_OS_WIN)
+    HRESULT hres;
+    IShellLink *psl;
+
+    // Get a pointer to the IShellLink interface. It is assumed that CoInitialize
+    // has already been called.
+    hres = CoCreateInstance(CLSID_ShellLink, NULL, CLSCTX_INPROC_SERVER, IID_IShellLink, (LPVOID *)&psl);
+    if (SUCCEEDED(hres)) {
+        IPersistFile *ppf;
+
+        // Set the path to the shortcut target and add the description.
+        WCHAR path[MAX_PATH];
+        GetModuleFileNameW(NULL, path, MAX_PATH);
+        psl->SetPath(path);
+        psl->SetDescription(L"Unipro UGENE");
+
+        // Query IShellLink for the IPersistFile interface, used for saving the
+        // shortcut in persistent storage.
+        hres = psl->QueryInterface(IID_IPersistFile, (LPVOID *)&ppf);
+
+        if (SUCCEEDED(hres)) {
+            WCHAR wsz[MAX_PATH + 1];
+            CHAR pathLink[MAX_PATH + 1];
+
+            if (SHGetSpecialFolderPathA(HWND_DESKTOP, pathLink, CSIDL_DESKTOP, FALSE)) {
+                strncat(pathLink, "\\UGENE.lnk", __min(strlen("\\UGENE.lnk"), MAX_PATH - strlen(pathLink)));
+
+                // Ensure that the string is Unicode.
+                MultiByteToWideChar(CP_ACP, 0, pathLink, -1, wsz, MAX_PATH);
+
+                // Add code here to check return value from MultiByteWideChar
+                // for success.
+
+                // Save the link by calling IPersistFile::Save.
+                hres = ppf->Save(wsz, TRUE);
+                ppf->Release();
+            } else {
+                return false;
+            }
+        }
+        psl->Release();
+    }
+    return SUCCEEDED(hres);
+#elif defined(Q_OS_LINUX)
+    QString homeDir = QDir::homePath();
+    QFile link(homeDir + "/Desktop/UGENE.desktop");
+    if (link.open(QFile::WriteOnly | QFile::Truncate)) {
+        QTextStream out(&link);
+        out.setCodec("UTF-8");
+        out << "[Desktop Entry]" << endl
+            << "Encoding=UTF-8" << endl
+            << "Version=1.0" << endl
+            << "Type=Application" << endl
+            << "Terminal=false" << endl
+            << "Exec=" + QCoreApplication::applicationFilePath() << endl
+            << "Name=Unipro UGENE" << endl
+            << "Icon=" + QCoreApplication::applicationDirPath() + "/ugene.png" << endl
+            << "Name[en_US]=Unipro UGENE" << endl;
+        link.close();
+        if (!link.setPermissions(link.permissions() | QFileDevice::ExeOwner | QFileDevice::ExeUser)) {
+            return false;
+        }
+        return true;
+    }
+    return false;
+#elif defined(Q_OS_MACX)
+    return false;
+#endif    // Q_OS_WIN
+}
+
+Task::ReportResult CreateDesktopShortcutTask::report() {

```

Comment:
I think, the the all code from this function could be moved to the "run" function. It's better, because "run" is in the separate thread.

Reply:
moved